MathLib/mechanics_invariants.h

```diff
@@ -0,0 +1,166 @@
+/**
+ * \copyright
+ * Copyright (c) 2012-2016, OpenGeoSys Community (http://www.opengeosys.org)
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ *
+ */
+#ifndef MATHLIB_MECHANICS_INVARIANTS_H
+#define MATHLIB_MECHANICS_INVARIANTS_H
+
+#include <Eigen/Dense>
+
+namespace MathLib
+{
+namespace Mechanics
+{
+/// Invariants used in mechanics, based on Kelvin representation of the vectors
+/// and matrices.
+/// The invariants are computed at process creation time.
+template <int KelvinVectorSize>
+struct Invariants final
+{
+    // Deviatoric projection matrix.
+    static Eigen::Matrix<double, KelvinVectorSize, KelvinVectorSize> const
+        P_dev;
+    // Spherical projection matrix.
+    static Eigen::Matrix<double, KelvinVectorSize, KelvinVectorSize> const
+        P_sph;
+    /// Kelvin mapping of 2nd order identity
+    static Eigen::Matrix<double, KelvinVectorSize, 1> const ivec;
+
+    // Determinant of a matrix in Kelvin vector representation.
+    static double determinant(
+        Eigen::Matrix<double, KelvinVectorSize, 1> const& vec);
+
+    // Inverse of a matrix in Kelvin vector representation.
+    static Eigen::Matrix<double, KelvinVectorSize, 1> inverse(
+        Eigen::Matrix<double, KelvinVectorSize, 1> const& vec);
+
+    static double effectiveStress(
+        Eigen::Matrix<double, KelvinVectorSize, 1> const& deviatoric_stress)
+    {
+        return std::sqrt(3 * J2(deviatoric_stress));
+    }
+
+    static double J2(
+        Eigen::Matrix<double, KelvinVectorSize, 1> const& deviatoric_stress)
+    {
+        return 0.5 * deviatoric_stress.transpose() * deviatoric_stress;
```

Comment:
The deviatoric part of a tensor is traceless (?), so there could be an assertion to check that.